### PR TITLE
Relaxed option naming

### PR DESCRIPTION
**Pull request by phlptp**
If merged this pull request will relax the restrictions on option naming and subcommand naming.  

Allowing inclusion of all characters except
starting with:

- `-` since we need to recognize `-` and `--` as different options and `--` has some special meaning
- `\n` since that would cause all sorts of chaos with the formatters and other things
- `!`  used as a special character for setting default values
- ` `  allowing this would create the potential for difficult to debug user entry errors for example `"-- input, input"` would be adding an extra space accidentally, But if space were allowed this would read fine as a single almost impossible to use option.  

and anywhere in the option
- `\n`  same as above
- `=`, and `:`  they are fine as a first character but after that they are used as a value separator so can't be allowed in an option
- `{`  `{}` is used to indicate default values
- ` `  allowing this would create the potential for difficult to debug user entry errors for example `"--input input"` would be forgetting a comma to separate an option from the positional.  But if space were allowed this would read fine as a single almost impossible to use option.  

this will address #580 
also allow lots of amusing things like `--:)`  as an option

A couple questions in #592 we talked about adding a check on group_names and a few other things.  particularly in regard to newlines.  I need to do that yet.  

But in regard to subcommands  through the `add_subcommand` I am inclined to leave the check for spaces in the subcommand name in place.  The reason has to do with error catching.    Most subcommand names don't have spaces but their description does quite frequently.  Meaning the check on spaces catches some potential errors in the API call.    Removing this restriction would likely cause additional errors that might be hard for a user to spot initially.    

Then the question is should we allow newlines in App names?    If not might be worth adding an additional validator to catch that and use it for app names, aliases, and group names.

### CODE REVIEW

**Comment by phlptp** [comment]
TODO
- [x] readme updates
- [x] additional test cases for subcommands
- [x] validators for group names, app name, aliases
- [x]  book updates

**Comment by phlptp** [comment]
Added a validator for group_names, aliases, and option group names.  But not app names.  App names can still have new lines, so it is possible to get a subcommand with a new line but not easy.    Should we restrict this as well?

File: book/chapters/flags.md [hunk 1]
```diff
@@ -120,4 +120,4 @@ Flag int: 3
 Flag plain: 1
 ```
 
-[^1]: It will not inherit this from the parent defaults, since this is often useful even if you don't want all options to allow multiple passed options.
+\[^1\]: It will not inherit this from the parent defaults, since this is often useful even if you don't want all options to allow multiple passed options.
```
Comment (henryiii) [inline book/chapters/flags.md:123]: Guessing this was required for some reason?

File: book/chapters/options.md [hunk 2]
```diff
@@ -26,12 +26,12 @@ You can use any C++ int-like type, not just `int`. CLI11 understands the followi
 | complex-number | std::complex or any type which has a real(), and imag() operations available, will allow 1 or 2 string definitions like "1+2j" or two arguments "1","2" |
 | enumeration | any enum or enum class type is supported through conversion from the underlying type(typically int, though it can be specified otherwise) |
 | container-like | a container(like vector) of any available types including other containers |
-| wrapper | any other object with a `value_type` static definition where the type specified by `value_type` is one of type in this list |
+| wrapper | any other object with a `value_type` static definition where the type specified by `value_type` is one of the type in this list, including std::atomic<>|
```
Comment (henryiii) [inline book/chapters/options.md]:
```suggestion
| wrapper | any other object with a `value_type` static definition where the type specified by `value_type` is one of the type in this list, including `std::atomic<>` |
```

File: include/CLI/App.hpp [hunk 3]
```diff
@@ -978,13 +967,13 @@ class App {
         if(!subcommand_name.empty() && !detail::valid_name_string(subcommand_name)) {
             if(!detail::valid_first_char(subcommand_name[0])) {
                 throw IncorrectConstruction(
-                    "Subcommand name starts with invalid character, allowed characters are [a-zA-z0-9]+'_','?','@' ");
+                    "Subcommand name starts with invalid character, '!' and'-' are not allowed");
```
Comment (henryiii) [inline include/CLI/App.hpp]:
```suggestion
                    "Subcommand name starts with invalid character, '!' and '-' are not allowed");
```

File: include/CLI/Option.hpp [hunk 4]
```diff
@@ -94,6 +94,9 @@ template <typename CRTP> class OptionBase {
 
     /// Changes the group membership
     CRTP *group(const std::string &name) {
+        if(!detail::valid_alias_name_string(name)) {
+            throw IncorrectConstruction("group names may not contain newlines or null characters");
```
Comment (henryiii) [inline include/CLI/Option.hpp]:
```suggestion
            throw IncorrectConstruction("Group names may not contain newlines or null characters");
```

File: book/chapters/flags.md (hunk 1 above)
Comment (phlptp) [reply book/chapters/flags.md:123]: In some parsers it interprets that as a link and give syntax warnings

**Pull request merged**